servicetalk-http-api/src/main/java/io/servicetalk/http/api/DefaultServiceDiscoveryRetryStrategy.java

```diff
@@ -0,0 +1,291 @@
+                                                 final BiIntFunction<Throwable, ? extends Completable> retryStrategy) {
+        this.retainTillReceivePercentage = retainTillReceivePercentage;
+        this.flipAvailability = requireNonNull(flipAvailability);
+        this.retryStrategy = requireNonNull(retryStrategy);
+    }
+
+    @Override
+    public Publisher<? extends E> apply(final Publisher<? extends E> sdEvents) {
+        return defer(() -> {
+            EventsCache<ResolvedAddress, E> eventsCache =
+                    new EventsCache<>(retainTillReceivePercentage, flipAvailability);
+            return sdEvents.flatMapConcatIterable(eventsCache::consume)
+                    .beforeOnError(__ -> eventsCache.errorSeen())
+                    .retryWhen(retryStrategy);
+        });
+    }
+
+    /**
+     * A builder to build instances of {@link DefaultServiceDiscoveryRetryStrategy}.
+     *
+     * @param <ResolvedAddress> The type of address after resolution.
+     * @param <E> Type of {@link ServiceDiscovererEvent}s published from {@link ServiceDiscoverer#discover(Object)}.
+     */
+    public static final class Builder<ResolvedAddress, E extends ServiceDiscovererEvent<ResolvedAddress>> {
+        // There is no reason for the choice of 75%, it is arbitrary and can be configured by users.
+        private int retainTillReceivePercentage = 75;
+        private BiIntFunction<Throwable, ? extends Completable> retryStrategy;
+        private final UnaryOperator<E> flipAvailability;
+
+        private Builder(final BiIntFunction<Throwable, ? extends Completable> retryStrategy,
+                        final UnaryOperator<E> flipAvailability) {
+            this.retryStrategy = retryStrategy;
+            this.flipAvailability = requireNonNull(flipAvailability);
+        }
+
+        /**
+         * A {@link Publisher} returned from {@link ServiceDiscoverer#discover(Object)} may fail transiently leaving the
+         * consumer of these events with an option of either disposing the addresses that were provided before the
+         * error or retain them till a retry succeeds. This option enables retention of addresses after an error is
+         * observed till the total number of received addresses after an error is {@code retainTillReceivePercentage}
+         * percent of the addresses that were available before the error was received.
+         *
+         * @param retainTillReceivePercentage A percentage of addresses that were received before an error that should
+         * be received after a success, after which unreceived addresses are removed by sending appropriate
+         * {@link ServiceDiscovererEvent}s. When set to {@code 0}, addresses are not retained.
+         * @return {@code this}.
+         */
+        public Builder<ResolvedAddress, E> retainAddressesTillSuccess(final int retainTillReceivePercentage) {
+            if (retainTillReceivePercentage < 0) {
```

Comment:
Minor ask : Requires upper cap as well retainTillReceivePercentage should not be greater than 100

Reply:
Not adding an upper cap is intentional here as it enables folks to retain older addresses for a longish time.